lib/io/can.hpp

```diff
@@ -1,12 +1,54 @@
 #pragma once
 
+#include <memory>
+#include <optional>
+#include <string>
+
 #include <core/types.hpp>
 
+#ifdef __linux__
+#include <linux/can.h>
+#endif
 namespace hyped::io {
 
-class Can {
+#ifdef __linux__
+using CanFrame = can_frame;
+#else
+// structs and values as defined in <linux/can.h>
+struct CanFrame {
+  std::uint32_t can_id;              // 32 bit CAN_ID + EFF/RTR/ERR flags
+  std::uint8_t can_dlc;              // frame payload length in bytes (0 .. CAN_MAX_DLEN)
+  std::uint8_t __pad;                // padding - do not asign value
+  std::uint8_t __res0;               // reserved - do not asign value
+  std::uint8_t __res1;               // reserved - do not asign value
+  std::array<std::uint8_t, 8> data;  // data that is sent over CAN, split into bytes
+};
+struct sockaddr_can {
+  std::uint16_t can_family;
+  int can_ifindex;
+  union {
+    // transport protocol class address info (e.g. ISOTP)
+    struct {
+      std::uint32_t rx_id, tx_id;
+    } tp;
+  } can_addr;
+};
+#define PF_CAN 29
+#define AF_CAN 29
+#define CAN_RAW 1
+#define CAN_MAX_DLEN 8
+#endif
+
+class ICanProcessor {
  public:
-  Can();
+  virtual void processMessage(const io::CanFrame &frame) = 0;
 };
 
-}  // namespace hyped::io
+class ICan {
+ public:
+  virtual core::Result send(const CanFrame &message)                                          = 0;
+  virtual std::optional<CanFrame> receive()                                                   = 0;
+  virtual core::Result listen()                                                               = 0;
```

Comment:
The name here isnt very descriptive, esp in the presence of receive.

I'm not even sure why we demand that an ICan instance implements both of these.

Also: What if theres more than one message waiting? Do we ever encounter a case where there's more messages coming in than we are collecting?

Reply:
the socket buffers some number of messages but i haven't determined an exact number, 3 <= n < 12000